### PR TITLE
Define write operations for status page API

### LINKED ISSUE #281: Extend status page OpenAPI spec to include write operations
The status page openapi spec needs to be extended to include the write operations for CUD operations.
This should follow the rest(-less) paradigm and work with proper http methods.

* partially to be retrieved from existing API server (https://github.com/sovereigncloudstack/status-page-api)
* leave out the auth features of openapi, document need for auth somewhere else

## Definition of Ready:

- [x] User Story is small enough to be finished within one sprint
- [x] User Story is clear and understood by the whole team
- [x] Acceptance criteria are defined
- [x] Acceptance criteria are clear and understood by the whole team

## Definition of Done:

- [x] The openapi spec includes the operations needed to create, update and delete ressources.
- [ ] All acceptance criteria are met
- [ ] Changes have been reviewed
- [ ] CI tests have run successfully
- [ ] Documentation has been updated
- [ ] Release Notes have been updated

### PR DESCRIPTION
**Pull request by joshuai96**
Define write operations for the status page Open API spec, mentioned in SovereignCloudStack/issues#281.

Includes:
- Create operations (`POST`) for:
  -  `PhaseList`
  - `ImpactType`
  - `Component`
  - `Incident`
  - `IncidentUpdate`
- Update opertations (`PATCH`) for:
  - `ImpactType`
  - `Component`
  - `Incident`
  - `IncidentUpdate`
- Delete operations (`DELETE`) for:
  -  `ImpactType`
  - `Components`
  - `Incident`
  - `IncidentUpdate`

`Phase`s are now handled as `PhaseList`, these use a generation system, to allow new lists to be created, for CUD operations on `Phases` without losing the reference.

The `IncidentUpdate`s are now handled as a real sub resource to incidents. 

`Impact`s are "helper" objects to connect `Incident`s and `Component`s and do not represent a standalone resource.

### CODE REVIEW

**Comment by joshuai96** [comment]
@joshmue I think it's now ready for the real review.

File: openapi.yaml [hunk 1]
```diff
@@ -4,93 +4,285 @@ info:
   version: '1.0'
 components:
   schemas:
+    # common values
     Id:
+      description: Identfication for objects. UUID prefered.
```
Comment (joshmue) [inline openapi.yaml]:
```suggestion
      description: Identification for objects. UUID preferred.
```

File: openapi.yaml [hunk 2]
```diff
@@ -4,93 +4,285 @@ info:
...
+      type: integer
+      readOnly: true
+
+    DisplayName:
+      description: Short and describing name.
       type: string
-    IncidentPhase:
+
+    Description:
+      description: A longer text for detailed informations.
```
Comment (joshmue) [inline openapi.yaml]:
```suggestion
      description: A longer text with detailed information.
```

File: openapi.yaml [hunk 3]
```diff
@@ -4,93 +4,285 @@ info:
...
       type: string
+
+    Date:
+      description: Point in time. Nullable for open ends.
```
Comment (joshmue) [inline openapi.yaml]:
```suggestion
      description: Point in time. Nullable for open end timeframes.
```

File: openapi.yaml [hunk 4]
```diff
@@ -4,93 +4,285 @@ info:
...
+        It's identified by the incident ID and its own order.
+        Updates happen in a given order.
       type: object
-      required: [text, createdAt]
       properties:
-        text:
-          type: string
+        order:
+          $ref: '#/components/schemas/Incremental'
+        displayName:
```
Comment (joshmue) [inline openapi.yaml:132]: An `IncidentUpdate` is having a `displayName`? Maybe rather `title` for `displayName` and `text` for `description`?

File: openapi.yaml [hunk 5]
```diff
@@ -4,93 +4,285 @@ info:
...
+          $ref: '#/components/schemas/DisplayName'
         description:
-          type: string
+          $ref: '#/components/schemas/Description'
         updates:
-          type: array
-          items:
-            $ref: '#/components/schemas/IncidentUpdate'
+          $ref: '#/components/schemas/Incremental'
```
Comment (joshmue) [inline openapi.yaml]: Rather `array` of `Incrementals`? Also maybe marking as `readOnly`, if possible?

File: openapi.yaml (hunk 4 above)
Comment (joshuai96) [reply openapi.yaml:132]: I just wanted to keep the field naming analogous to the other objects. A display name is just a name to display. A title could be seen as a subtype of a display name.

Comment (joshmue) [reply openapi.yaml:132]: Keeping it as-is works for me :+1:

**Review by joshmue — COMMENTED** [review]
Noticed that when I looked at the spec in Swagger UI:

Some routes like `POST /components` return bare JSON strings, some routes like `PATCH /components/:id` return bare JSON booleans. Also, some routes like `GET /components` return unwrapped JSON arrays.

While this works, this may bring maintainability issues and/or security issues like JSON hijacking.

IIRC, for reasons like these, wrapping response data in JSON objects is common practice in most API's; We should probably do that as well.

**Comment by joshuai96** [comment]
@joshmue all responses should be wrapped now.

**Comment by joshmue** [comment]
Now, the spec allows returning pretty much anything at any endpoint. For example, returning an `Incident` at the `GET /impacttypes` endpoint is okay according to the spec. This has at least the following implications:

- the Swagger UI shows `{"content": true}` as example response for e. g. `GET /impacttypes`
- (probably) generated client code will require the user to do additional type assertions/casting/... to get to the actual data

The first point could be tackled by defining explicit examples in the spec, but that would not help with the second one.

The best option may be to define more explicit responses in `responses`, I guess.

File: docs/component_overview.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# Component Overview
+
+This represents a part of the decission process related to the overall structure the API wants to represent.
```
Comment (joshmue) [inline docs/component_overview.md]:
```suggestion
This represents a part of the decision process related to the overall structure the API wants to represent.
```

**Review by joshmue — APPROVED** [review]
LGTM!

**Pull request merged**